# This file is part of HiGFXback

# requires
REQUIRES="autotools-wrappers-lt-build curl-build gst-plugins-base-build libxml2-build libwebp-build openssl-build"

PKG_CONFIG_PATH=/fb/share/pkgconfig pkg-config --exists --print-errors $REQUIRES || exit 1

if pkg-config --exists openjpeg2-build; then
  REQUIRES="$REQUIRES openjpeg2-build"
else
  pkg-config --exists --print-errors openjpeg-build || exit 1
  REQUIRES="$REQUIRES openjpeg-build"
fi

if PKG_CONFIG_PATH=/fb/share/pkgconfig pkg-config --exists libde265-build; then
  DE265=1
  REQUIRES="$REQUIRES libde265-build"
fi

if pkg-config --exists openal-soft-build; then
  OPENAL=1
  REQUIRES="$REQUIRES openal-soft-build"
fi

if pkg-config --exists x265-build; then
  X265=1
  REQUIRES="$REQUIRES x265-build"
fi

# configure (ac-2.69; am-1.15; lt-2.4.2)
libtoolize -c -f; aclocal -I m4 -I common/m4; autoconf; autoheader -f; automake -a -c
PKG_CONFIG_PATH=/fb/lib/pkgconfig LDFLAGS=-Wl,-rpath,/fb/lib ./configure --disable-decklink --prefix=/fb

# build
make

# install
install -d $DESTDIR/fb/lib
install gst-libs/gst/adaptivedemux/.libs/libgstadaptivedemux-1.0.so.0.* $DESTDIR/fb/lib/libgstadaptivedemux-1.0.so.0
install gst-libs/gst/base/.libs/libgstbadbase-1.0.so.0.* $DESTDIR/fb/lib/libgstbadbase-1.0.so.0
install gst-libs/gst/basecamerabinsrc/.libs/libgstbasecamerabinsrc-1.0.so.0.* $DESTDIR/fb/lib/libgstbasecamerabinsrc-1.0.so.0
install gst-libs/gst/codecparsers/.libs/libgstcodecparsers-1.0.so.0.* $DESTDIR/fb/lib/libgstcodecparsers-1.0.so.0
install gst-libs/gst/interfaces/.libs/libgstphotography-1.0.so.0.* $DESTDIR/fb/lib/libgstphotography-1.0.so.0
install gst-libs/gst/mpegts/.libs/libgstmpegts-1.0.so.0.* $DESTDIR/fb/lib/libgstmpegts-1.0.so.0
install gst-libs/gst/uridownloader/.libs/libgsturidownloader-1.0.so.0.* $DESTDIR/fb/lib/libgsturidownloader-1.0.so.0
install gst-libs/gst/video/.libs/libgstbadvideo-1.0.so.0.* $DESTDIR/fb/lib/libgstbadvideo-1.0.so.0
install -d $DESTDIR/fb/lib/gstreamer-1.0
install ext/bz2/.libs/libgstbz2.so $DESTDIR/fb/lib/gstreamer-1.0
install ext/curl/.libs/libgstcurl.so $DESTDIR/fb/lib/gstreamer-1.0
install ext/dash/.libs/libgstdashdemux.so $DESTDIR/fb/lib/gstreamer-1.0
install ext/dtls/.libs/libgstdtls.so $DESTDIR/fb/lib/gstreamer-1.0
install ext/hls/.libs/libgstfragmented.so $DESTDIR/fb/lib/gstreamer-1.0
test $DE265 && install ext/libde265/.libs/libgstlibde265.so $DESTDIR/fb/lib/gstreamer-1.0
test $OPENAL && install ext/openal/.libs/libgstopenal.so $DESTDIR/fb/lib/gstreamer-1.0
install ext/openjpeg/.libs/libgstopenjpeg.so $DESTDIR/fb/lib/gstreamer-1.0
install ext/smoothstreaming/.libs/libgstsmoothstreaming.so $DESTDIR/fb/lib/gstreamer-1.0
install ext/webp/.libs/libgstwebp.so $DESTDIR/fb/lib/gstreamer-1.0
test $X265 && install ext/x265/.libs/libgstx265.so $DESTDIR/fb/lib/gstreamer-1.0
install gst/accurip/.libs/libgstaccurip.so $DESTDIR/fb/lib/gstreamer-1.0
install gst/adpcmdec/.libs/libgstadpcmdec.so $DESTDIR/fb/lib/gstreamer-1.0
install gst/adpcmenc/.libs/libgstadpcmenc.so $DESTDIR/fb/lib/gstreamer-1.0
install gst/aiff/.libs/libgstaiff.so $DESTDIR/fb/lib/gstreamer-1.0
install gst/asfmux/.libs/libgstasfmux.so $DESTDIR/fb/lib/gstreamer-1.0
install gst/audiofxbad/.libs/libgstaudiofxbad.so $DESTDIR/fb/lib/gstreamer-1.0
install gst/audiomixer/.libs/libgstaudiomixer.so $DESTDIR/fb/lib/gstreamer-1.0
install gst/audiovisualizers/.libs/libgstaudiovisualizers.so $DESTDIR/fb/lib/gstreamer-1.0
install gst/autoconvert/.libs/libgstautoconvert.so $DESTDIR/fb/lib/gstreamer-1.0
install gst/bayer/.libs/libgstbayer.so $DESTDIR/fb/lib/gstreamer-1.0
install gst/camerabin2/.libs/libgstcamerabin2.so $DESTDIR/fb/lib/gstreamer-1.0
install gst/coloreffects/.libs/libgstcoloreffects.so $DESTDIR/fb/lib/gstreamer-1.0
install gst/compositor/.libs/libgstcompositor.so $DESTDIR/fb/lib/gstreamer-1.0
install gst/dataurisrc/.libs/libgstdataurisrc.so $DESTDIR/fb/lib/gstreamer-1.0
install gst/debugutils/.libs/libgstdebugutilsbad.so $DESTDIR/fb/lib/gstreamer-1.0
install gst/dvbsuboverlay/.libs/libgstdvbsuboverlay.so $DESTDIR/fb/lib/gstreamer-1.0
install gst/dvdspu/.libs/libgstdvdspu.so $DESTDIR/fb/lib/gstreamer-1.0
install gst/festival/.libs/libgstfestival.so $DESTDIR/fb/lib/gstreamer-1.0
install gst/fieldanalysis/.libs/libgstfieldanalysis.so $DESTDIR/fb/lib/gstreamer-1.0
install gst/freeverb/.libs/libgstfreeverb.so $DESTDIR/fb/lib/gstreamer-1.0
install gst/frei0r/.libs/libgstfrei0r.so $DESTDIR/fb/lib/gstreamer-1.0
install gst/gaudieffects/.libs/libgstgaudieffects.so $DESTDIR/fb/lib/gstreamer-1.0
install gst/gdp/.libs/libgstgdp.so $DESTDIR/fb/lib/gstreamer-1.0
install gst/geometrictransform/.libs/libgstgeometrictransform.so $DESTDIR/fb/lib/gstreamer-1.0
install gst/id3tag/.libs/libgstid3tag.so $DESTDIR/fb/lib/gstreamer-1.0
install gst/inter/.libs/libgstinter.so $DESTDIR/fb/lib/gstreamer-1.0
install gst/interlace/.libs/libgstinterlace.so $DESTDIR/fb/lib/gstreamer-1.0
install gst/ivfparse/.libs/libgstivfparse.so $DESTDIR/fb/lib/gstreamer-1.0
install gst/ivtc/.libs/libgstivtc.so $DESTDIR/fb/lib/gstreamer-1.0
install gst/jp2kdecimator/.libs/libgstjp2kdecimator.so $DESTDIR/fb/lib/gstreamer-1.0
install gst/jpegformat/.libs/libgstjpegformat.so $DESTDIR/fb/lib/gstreamer-1.0
install gst/librfb/.libs/libgstrfbsrc.so $DESTDIR/fb/lib/gstreamer-1.0
install gst/liveadder/.libs/libgstliveadder.so $DESTDIR/fb/lib/gstreamer-1.0
install gst/midi/.libs/libgstmidi.so $DESTDIR/fb/lib/gstreamer-1.0
install gst/mpegdemux/.libs/libgstmpegpsdemux.so $DESTDIR/fb/lib/gstreamer-1.0
install gst/mpegpsmux/.libs/libgstmpegpsmux.so $DESTDIR/fb/lib/gstreamer-1.0
install gst/mpegtsdemux/.libs/libgstmpegtsdemux.so $DESTDIR/fb/lib/gstreamer-1.0
install gst/mpegtsmux/.libs/libgstmpegtsmux.so $DESTDIR/fb/lib/gstreamer-1.0
install gst/mxf/.libs/libgstmxf.so $DESTDIR/fb/lib/gstreamer-1.0
install gst/pcapparse/.libs/libgstpcapparse.so $DESTDIR/fb/lib/gstreamer-1.0
install gst/pnm/.libs/libgstpnm.so $DESTDIR/fb/lib/gstreamer-1.0
install gst/rawparse/.libs/libgstrawparse.so $DESTDIR/fb/lib/gstreamer-1.0
install gst/removesilence/.libs/libgstremovesilence.so $DESTDIR/fb/lib/gstreamer-1.0
install gst/rtp/.libs/libgstrtpbad.so $DESTDIR/fb/lib/gstreamer-1.0
install gst/onvif/.libs/libgstrtponvif.so $DESTDIR/fb/lib/gstreamer-1.0
install gst/sdp/.libs/libgstsdpelem.so $DESTDIR/fb/lib/gstreamer-1.0
install gst/segmentclip/.libs/libgstsegmentclip.so $DESTDIR/fb/lib/gstreamer-1.0
install gst/siren/.libs/libgstsiren.so $DESTDIR/fb/lib/gstreamer-1.0
install gst/smooth/.libs/libgstsmooth.so $DESTDIR/fb/lib/gstreamer-1.0
install gst/speed/.libs/libgstspeed.so $DESTDIR/fb/lib/gstreamer-1.0
install gst/stereo/.libs/libgststereo.so $DESTDIR/fb/lib/gstreamer-1.0
install gst/subenc/.libs/libgstsubenc.so $DESTDIR/fb/lib/gstreamer-1.0
install gst/videofilters/.libs/libgstvideofiltersbad.so $DESTDIR/fb/lib/gstreamer-1.0
install gst/videoparsers/.libs/libgstvideoparsersbad.so $DESTDIR/fb/lib/gstreamer-1.0
install gst/videosignal/.libs/libgstvideosignal.so $DESTDIR/fb/lib/gstreamer-1.0
install gst/vmnc/.libs/libgstvmnc.so $DESTDIR/fb/lib/gstreamer-1.0
install gst/y4m/.libs/libgsty4mdec.so $DESTDIR/fb/lib/gstreamer-1.0
install gst/yadif/.libs/libgstyadif.so $DESTDIR/fb/lib/gstreamer-1.0
install sys/dvb/.libs/libgstdvb.so $DESTDIR/fb/lib/gstreamer-1.0
install sys/fbdev/.libs/libgstfbdevsink.so $DESTDIR/fb/lib/gstreamer-1.0
install sys/shm/.libs/libgstshm.so $DESTDIR/fb/lib/gstreamer-1.0
install sys/vcd/.libs/libgstvcdsrc.so $DESTDIR/fb/lib/gstreamer-1.0

# build.pc
install -d $DESTDIR/fb/share/pkgconfig
cat > $DESTDIR/fb/share/pkgconfig/gst-plugins-bad-build.pc << EOF
Name: gst-plugins-bad
Version: 1.6.4
Description: Bad GStreamer plugins
Requires: $REQUIRES

exec=\\
/fb/lib/libgstadaptivedemux-1.0.so.0 \\
/fb/lib/libgstbadbase-1.0.so.0 \\
/fb/lib/libgstbadvideo-1.0.so.0 \\
/fb/lib/libgstbasecamerabinsrc-1.0.so.0 \\
/fb/lib/libgstcodecparsers-1.0.so.0 \\
/fb/lib/libgstmpegts-1.0.so.0 \\
/fb/lib/libgstphotography-1.0.so.0 \\
/fb/lib/libgsturidownloader-1.0.so.0 \\
/fb/lib/gstreamer-1.0/libgstaccurip.so \\
/fb/lib/gstreamer-1.0/libgstadpcmdec.so \\
/fb/lib/gstreamer-1.0/libgstadpcmenc.so \\
/fb/lib/gstreamer-1.0/libgstaiff.so \\
/fb/lib/gstreamer-1.0/libgstasfmux.so \\
/fb/lib/gstreamer-1.0/libgstaudiofxbad.so \\
/fb/lib/gstreamer-1.0/libgstaudiomixer.so \\
/fb/lib/gstreamer-1.0/libgstaudiovisualizers.so \\
/fb/lib/gstreamer-1.0/libgstautoconvert.so \\
/fb/lib/gstreamer-1.0/libgstbayer.so \\
/fb/lib/gstreamer-1.0/libgstbz2.so \\
/fb/lib/gstreamer-1.0/libgstcamerabin2.so \\
/fb/lib/gstreamer-1.0/libgstcoloreffects.so \\
/fb/lib/gstreamer-1.0/libgstcompositor.so \\
/fb/lib/gstreamer-1.0/libgstcurl.so \\
/fb/lib/gstreamer-1.0/libgstdashdemux.so \\
/fb/lib/gstreamer-1.0/libgstdataurisrc.so \\
/fb/lib/gstreamer-1.0/libgstdebugutilsbad.so \\
/fb/lib/gstreamer-1.0/libgstdtls.so \\
/fb/lib/gstreamer-1.0/libgstdvb.so \\
/fb/lib/gstreamer-1.0/libgstdvbsuboverlay.so \\
/fb/lib/gstreamer-1.0/libgstdvdspu.so \\
/fb/lib/gstreamer-1.0/libgstfbdevsink.so \\
/fb/lib/gstreamer-1.0/libgstfestival.so \\
/fb/lib/gstreamer-1.0/libgstfieldanalysis.so \\
/fb/lib/gstreamer-1.0/libgstfragmented.so \\
/fb/lib/gstreamer-1.0/libgstfreeverb.so \\
/fb/lib/gstreamer-1.0/libgstfrei0r.so \\
/fb/lib/gstreamer-1.0/libgstgaudieffects.so \\
/fb/lib/gstreamer-1.0/libgstgdp.so \\
/fb/lib/gstreamer-1.0/libgstgeometrictransform.so \\
/fb/lib/gstreamer-1.0/libgstid3tag.so \\
/fb/lib/gstreamer-1.0/libgstinter.so \\
/fb/lib/gstreamer-1.0/libgstinterlace.so \\
/fb/lib/gstreamer-1.0/libgstivfparse.so \\
/fb/lib/gstreamer-1.0/libgstivtc.so \\
/fb/lib/gstreamer-1.0/libgstjp2kdecimator.so \\
/fb/lib/gstreamer-1.0/libgstjpegformat.so \\
EOF
test $DE265 && echo /fb/lib/gstreamer-1.0/libgstlibde265.so \\ >> $DESTDIR/fb/share/pkgconfig/gst-plugins-bad-build.pc
cat >> $DESTDIR/fb/share/pkgconfig/gst-plugins-bad-build.pc << EOF
/fb/lib/gstreamer-1.0/libgstliveadder.so \\
/fb/lib/gstreamer-1.0/libgstmidi.so \\
/fb/lib/gstreamer-1.0/libgstmpegpsdemux.so \\
/fb/lib/gstreamer-1.0/libgstmpegpsmux.so \\
/fb/lib/gstreamer-1.0/libgstmpegtsdemux.so \\
/fb/lib/gstreamer-1.0/libgstmpegtsmux.so \\
/fb/lib/gstreamer-1.0/libgstmxf.so \\
EOF
test $OPENAL && echo /fb/lib/gstreamer-1.0/libgstopenal.so \\ >> $DESTDIR/fb/share/pkgconfig/gst-plugins-bad-build.pc
cat >> $DESTDIR/fb/share/pkgconfig/gst-plugins-bad-build.pc << EOF
/fb/lib/gstreamer-1.0/libgstopenjpeg.so \\
/fb/lib/gstreamer-1.0/libgstpcapparse.so \\
/fb/lib/gstreamer-1.0/libgstpnm.so \\
/fb/lib/gstreamer-1.0/libgstrawparse.so \\
/fb/lib/gstreamer-1.0/libgstremovesilence.so \\
/fb/lib/gstreamer-1.0/libgstrfbsrc.so \\
/fb/lib/gstreamer-1.0/libgstrtpbad.so \\
/fb/lib/gstreamer-1.0/libgstrtponvif.so \\
/fb/lib/gstreamer-1.0/libgstsdpelem.so \\
/fb/lib/gstreamer-1.0/libgstsegmentclip.so \\
/fb/lib/gstreamer-1.0/libgstshm.so \\
/fb/lib/gstreamer-1.0/libgstsiren.so \\
/fb/lib/gstreamer-1.0/libgstsmooth.so \\
/fb/lib/gstreamer-1.0/libgstsmoothstreaming.so \\
/fb/lib/gstreamer-1.0/libgstspeed.so \\
/fb/lib/gstreamer-1.0/libgststereo.so \\
/fb/lib/gstreamer-1.0/libgstsubenc.so \\
/fb/lib/gstreamer-1.0/libgstvcdsrc.so \\
/fb/lib/gstreamer-1.0/libgstvideofiltersbad.so \\
/fb/lib/gstreamer-1.0/libgstvideoparsersbad.so \\
/fb/lib/gstreamer-1.0/libgstvideosignal.so \\
/fb/lib/gstreamer-1.0/libgstvmnc.so \\
/fb/lib/gstreamer-1.0/libgstwebp.so \\
EOF
test $X265 && echo /fb/lib/gstreamer-1.0/libgstx265.so \\ >> $DESTDIR/fb/share/pkgconfig/gst-plugins-bad-build.pc
cat >> $DESTDIR/fb/share/pkgconfig/gst-plugins-bad-build.pc << EOF
/fb/lib/gstreamer-1.0/libgsty4mdec.so \\
/fb/lib/gstreamer-1.0/libgstyadif.so
EOF

exit
--- gst-plugins-bad-1.6.4.orig/ext/dash/gstdashdemux.c
+++ gst-plugins-bad-1.6.4/ext/dash/gstdashdemux.c
@@ -150,7 +150,7 @@
 #include <gio/gio.h>
 #include <gst/base/gsttypefindhelper.h>
 #include <gst/tag/tag.h>
-#include <gst/net/gstnet.h>
+#include <gst/net/net.h>
 #include "gst/gst-i18n-plugin.h"
 #include "gstdashdemux.h"
 #include "gstdash_debug.h"
--- gst-plugins-bad-1.6.4.orig/ext/dtls/gstdtlsconnection.c
+++ gst-plugins-bad-1.6.4/ext/dtls/gstdtlsconnection.c
@@ -239,7 +239,7 @@
       priv->bio = BIO_new (BIO_s_gst_dtls_connection ());
       g_return_if_fail (priv->bio);
 
-      priv->bio->ptr = self;
+      BIO_set_data (priv->bio, self);
       SSL_set_bio (priv->ssl, priv->bio, priv->bio);
 
       SSL_set_verify (priv->ssl,
@@ -573,10 +573,9 @@
   states |= (! !SSL_want_write (priv->ssl) << 20);
   states |= (! !SSL_want_read (priv->ssl) << 24);
 
-  GST_LOG_OBJECT (self, "%s: role=%s buf=(%d,%p:%d/%d) %x|%x %s",
+  GST_LOG_OBJECT (self, "%s: role=%s buf=(%p:%d/%d) %x|%x %s",
       str,
       priv->is_client ? "client" : "server",
-      pqueue_size (priv->ssl->d1->sent_messages),
       priv->bio_buffer,
       priv->bio_buffer_offset,
       priv->bio_buffer_len,
@@ -737,7 +736,7 @@
   self = SSL_get_ex_data (ssl, connection_ex_index);
   g_return_val_if_fail (GST_IS_DTLS_CONNECTION (self), FALSE);
 
-  pem = _gst_dtls_x509_to_pem (x509_ctx->cert);
+  pem = _gst_dtls_x509_to_pem (X509_STORE_CTX_get0_cert (x509_ctx));
 
   if (!pem) {
     GST_WARNING_OBJECT (self,
@@ -749,7 +748,8 @@
       gint len;
 
       len =
-          X509_NAME_print_ex (bio, X509_get_subject_name (x509_ctx->cert), 1,
+          X509_NAME_print_ex (bio,
+          X509_get_subject_name (X509_STORE_CTX_get0_cert (x509_ctx)), 1,
           XN_FLAG_MULTILINE);
       BIO_read (bio, buffer, len);
       buffer[len] = '\0';
@@ -777,29 +777,32 @@
     ########  ####  #######
 */
 
-static BIO_METHOD custom_bio_methods = {
-  BIO_TYPE_BIO,
-  "stream",
-  bio_method_write,
-  bio_method_read,
-  NULL,
-  NULL,
-  bio_method_ctrl,
-  bio_method_new,
-  bio_method_free,
-  NULL,
-};
+static BIO_METHOD *custom_bio_methods;
 
 static BIO_METHOD *
 BIO_s_gst_dtls_connection (void)
 {
-  return &custom_bio_methods;
+  if (custom_bio_methods != NULL)
+    return custom_bio_methods;
+
+  custom_bio_methods = BIO_meth_new (BIO_TYPE_BIO, "stream");
+  if (custom_bio_methods == NULL
+      || !BIO_meth_set_write (custom_bio_methods, bio_method_write)
+      || !BIO_meth_set_read (custom_bio_methods, bio_method_read)
+      || !BIO_meth_set_ctrl (custom_bio_methods, bio_method_ctrl)
+      || !BIO_meth_set_create (custom_bio_methods, bio_method_new)
+      || !BIO_meth_set_destroy (custom_bio_methods, bio_method_free)) {
+    BIO_meth_free (custom_bio_methods);
+    return NULL;
+  }
+
+  return custom_bio_methods;
 }
 
 static int
 bio_method_write (BIO * bio, const char *data, int size)
 {
-  GstDtlsConnection *self = GST_DTLS_CONNECTION (bio->ptr);
+  GstDtlsConnection *self = GST_DTLS_CONNECTION (BIO_get_data (bio));
 
   GST_LOG_OBJECT (self, "BIO: writing %d", size);
 
@@ -824,7 +827,7 @@
 static int
 bio_method_read (BIO * bio, char *out_buffer, int size)
 {
-  GstDtlsConnection *self = GST_DTLS_CONNECTION (bio->ptr);
+  GstDtlsConnection *self = GST_DTLS_CONNECTION (BIO_get_data (bio));
   GstDtlsConnectionPrivate *priv = self->priv;
   guint internal_size;
   gint copy_size;
@@ -868,7 +871,7 @@
 static long
 bio_method_ctrl (BIO * bio, int cmd, long arg1, void *arg2)
 {
-  GstDtlsConnection *self = GST_DTLS_CONNECTION (bio->ptr);
+  GstDtlsConnection *self = GST_DTLS_CONNECTION (BIO_get_data (bio));
   GstDtlsConnectionPrivate *priv = self->priv;
 
   switch (cmd) {
@@ -916,8 +919,8 @@
 {
   GST_LOG_OBJECT (NULL, "BIO: new");
 
-  bio->shutdown = 0;
-  bio->init = 1;
+  BIO_set_shutdown (bio, 0);
+  BIO_set_init (bio, 1);
 
   return 1;
 }
@@ -930,6 +933,6 @@
     return 0;
   }
 
-  GST_LOG_OBJECT (GST_DTLS_CONNECTION (bio->ptr), "BIO free");
+  GST_LOG_OBJECT (GST_DTLS_CONNECTION (BIO_get_data (bio)), "BIO free");
   return 0;
 }
--- gst-plugins-bad-1.6.4.orig/ext/hls/gsthlsdemux.c
+++ gst-plugins-bad-1.6.4/ext/hls/gsthlsdemux.c
@@ -1113,11 +1113,11 @@
 gst_hls_demux_decrypt_start (GstHLSDemux * demux, const guint8 * key_data,
     const guint8 * iv_data)
 {
-  EVP_CIPHER_CTX_init (&demux->aes_ctx);
-  if (!EVP_DecryptInit_ex (&demux->aes_ctx, EVP_aes_128_cbc (), NULL, key_data,
+  demux->aes_ctx = EVP_CIPHER_CTX_new ();
+  if (!EVP_DecryptInit_ex (demux->aes_ctx, EVP_aes_128_cbc (), NULL, key_data,
           iv_data))
     return FALSE;
-  EVP_CIPHER_CTX_set_padding (&demux->aes_ctx, 0);
+  EVP_CIPHER_CTX_set_padding (demux->aes_ctx, 0);
   return TRUE;
 }
 
@@ -1131,10 +1131,10 @@
     return FALSE;
 
   len = (int) length;
-  if (!EVP_DecryptUpdate (&demux->aes_ctx, decrypted_data, &len, encrypted_data,
+  if (!EVP_DecryptUpdate (demux->aes_ctx, decrypted_data, &len, encrypted_data,
           len))
     return FALSE;
-  EVP_DecryptFinal_ex (&demux->aes_ctx, decrypted_data + len, &flen);
+  EVP_DecryptFinal_ex (demux->aes_ctx, decrypted_data + len, &flen);
   g_return_val_if_fail (len + flen == length, FALSE);
   return TRUE;
 }
@@ -1142,7 +1142,8 @@
 static void
 gst_hls_demux_decrypt_end (GstHLSDemux * demux)
 {
-  EVP_CIPHER_CTX_cleanup (&demux->aes_ctx);
+  EVP_CIPHER_CTX_free (demux->aes_ctx);
+  demux->aes_ctx = NULL;
 }
 
 #elif defined(HAVE_NETTLE)
--- gst-plugins-bad-1.6.4.orig/ext/hls/gsthlsdemux.h
+++ gst-plugins-bad-1.6.4/ext/hls/gsthlsdemux.h
@@ -76,7 +76,7 @@
 
   /* decryption tooling */
 #if defined(HAVE_OPENSSL)
-  EVP_CIPHER_CTX aes_ctx;
+  EVP_CIPHER_CTX *aes_ctx;
 #elif defined(HAVE_NETTLE)
   struct CBC_CTX (struct aes_ctx, AES_BLOCK_SIZE) aes_ctx;
 #else
--- gst-plugins-bad-1.6.4.orig/gst-libs/gst/basecamerabinsrc/gstcamerabinpreview.c
+++ gst-plugins-bad-1.6.4/gst-libs/gst/basecamerabinsrc/gstcamerabinpreview.c
@@ -26,8 +26,7 @@
  * #GstCameraBinVideo.
  *
  */
-#include <gst/app/gstappsrc.h>
-#include <gst/app/gstappsink.h>
+#include <gst/app/app.h>
 #include <gst/glib-compat-private.h>
 #include "gstcamerabinpreview.h"
 #include "gstbasecamerasrc.h"
